### PR TITLE
Implement Harmonic Mean and add tests

### LINKED ISSUE #2035: Add `tfa.metrics.HarmonicMean`
**Describe the feature and the current behavior/state.**
Computes HarmonicMean

**Relevant information**
- Are you willing to contribute it (yes/no): YES
- Are you willing to maintain it going forward? (yes/no): YES
- Is there a relevant academic paper? (if so, where): NO
- Is there already an implementation in another framework? (if so, where): python stdlib statistics
- Was it part of tf.contrib? (if so, where): NO

**Which API type would this fall under (layer, metric, optimizer, etc.)**
`tfa.metrics`

**Who will benefit with this feature?**
Developers and researchers who want to obtain representative values using harmonic mean

**Any other info.**

### PR DESCRIPTION
**Pull request by rushabh-v**
# Description

Brief Description of the PR:

Fixes #2035 

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [x] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [x] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

Note: I have manually confirmed the correctness of the tests.

### CODE REVIEW

**Comment by bhack** [comment]
Thanks for the PR, 
honestly as it is just a little bit more than:
```
m2 =tf.keras.metrics.Mean()
m2.update_state(tf.math.reciprocal(input))
print(tf.math.reciprocal(m2.result()).numpy())
```

I suggest that it could be added with an extra reduction case in a TF core PR for `result`:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L427-L437

Plus an extra `enum` at: 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/utils/metrics_utils.py#L48-L60

And a single line code for a new Class `HarmonicMean` at:
https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/python/keras/metrics.py#L445-L483

**Comment by bhack** [comment]
Probably you can also simply ovverride in TF:

```python
import tensorflow as tf
from scipy.stats import hmean
from tensorflow.python.keras import metrics, utils

class HarmonicMean(metrics.Reduce):
  def __init__(self, name='mean', dtype=None):
    super(HarmonicMean, self).__init__(
        reduction=utils.metrics_utils.Reduction.WEIGHTED_MEAN, name=name, dtype=dtype)
  def result(self):
    return tf.math.reciprocal(tf.math.divide_no_nan(self.total, self.count))
  def update_state(self, values, sample_weight=None):
      values = tf.math.reciprocal(values)
      super().update_state(values, sample_weight)

m2 =HarmonicMean()
m2.update_state(input)
print(m2.result().numpy())

print(hmean([1, 3, 5, 7]))
```

**Comment by rushabh-v** [comment]
Okay, Sounds fair!
Should I close this PR and open one in TF-Core?

**Comment by bhack** [comment]
Yes, please keep this open untill the new one is accepted in TF so that we have a backup solution to work on.

**Comment by bhack** [comment]
Ok after the ecosystem review It seems that we could implement this here instead of TF core.

Can we derive from public Class `tf.keras.metrics.Mean` and override only what we need?

**Comment by rushabh-v** [comment]
Okay, sure!

**Comment by rushabh-v** [comment]
The failing check says `ConnectionResetError(104, 'Connection reset by peer'))`

File: tensorflow_addons/metrics/harmonic_mean.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    @typechecked
+    def __init__(
+        self, name: str = "harmonic_mean", dtype: AcceptableDTypes = None, **kwargs
+    ):
+        super().__init__(name=name, dtype=dtype, **kwargs)
+
+    def update_state(self, values, sample_weight=None) -> None:
+        values = tf.cast(values, dtype=self.dtype)
+        sample_weight = sample_weight_shape_match(values, sample_weight)
```
Comment (bhack) [inline tensorflow_addons/metrics/harmonic_mean.py:48]: Is this not part of the parent update?

File: tensorflow_addons/metrics/harmonic_mean.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        sample_weight = tf.cast(sample_weight, dtype=self.dtype)
+        super().update_state(tf.math.reciprocal(values), sample_weight)
```
Comment (bhack) [inline tensorflow_addons/metrics/harmonic_mean.py:50]: Do we need to filter with `reciprocal_nan` here or mask `tf.boolean_mask(x, tf.math.is_finite(values))`?

File: tensorflow_addons/metrics/harmonic_mean.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    def result(self) -> tf.Tensor:
+        ret = tf.math.reciprocal(super().result()).numpy()
```
Comment (bhack) [inline tensorflow_addons/metrics/harmonic_mean.py:53]: Ditto

File: tensorflow_addons/metrics/tests/harmonic_mean_test.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+"""Tests HarmonicMean metrics."""
+
+import numpy as np
+import pytest
+import tensorflow as tf
+
+from tensorflow_addons.metrics import HarmonicMean
+
+
+def get_test_data():
```
Comment (bhack) [inline tensorflow_addons/metrics/tests/harmonic_mean_test.py:24]: As the base class is covered by https://github.com/tensorflow/tensorflow/blob/v2.4.0/tensorflow/python/keras/metrics_test.py#L169 what we really need to test here for the overrided methods?

**Review by bhack — CHANGES_REQUESTED** [review]

File: tensorflow_addons/metrics/harmonic_mean.py (hunk 1 above)
Comment (rushabh-v) [reply tensorflow_addons/metrics/harmonic_mean.py:48]: No it raises `NotFoundError: Could not find device for node: {{node Reciprocal}} = Reciprocal[T=DT_INT32]` without that.

File: tensorflow_addons/metrics/harmonic_mean.py (hunk 2 above)
Comment (rushabh-v) [reply tensorflow_addons/metrics/harmonic_mean.py:50]: Do you mean `reciprocal_no_nan`?

Assuming that, yeah I think that would make more sense.

File: tensorflow_addons/metrics/tests/harmonic_mean_test.py (hunk 4 above)
Comment (rushabh-v) [reply tensorflow_addons/metrics/tests/harmonic_mean_test.py:24]: Okay, I'll remove all the extra tests (config tests, state tests, etc.)

File: tensorflow_addons/metrics/harmonic_mean.py (hunk 1 above)
Comment (bhack) [reply tensorflow_addons/metrics/harmonic_mean.py:48]: On the `sample_weights` or `values`?

File: tensorflow_addons/metrics/harmonic_mean.py (hunk 2 above)
Comment (bhack) [reply tensorflow_addons/metrics/harmonic_mean.py:50]: Yes and masking for infinite

Comment (rushabh-v) [reply tensorflow_addons/metrics/harmonic_mean.py:50]: Looking at the `scipy.stats.hmean's` behavior, infinite values should be allowed IMO. And about using `tf.reciprocal_no_nan`, I realized that doing that would turn into unexpected results when there are 0's in the input data.

File: tensorflow_addons/metrics/harmonic_mean.py (hunk 1 above)
Comment (rushabh-v) [reply tensorflow_addons/metrics/harmonic_mean.py:48]: Oh, sorry I missed this message.

> On the sample_weights or values?

On `values`. Calling `reciprocal` on Integer type data raises this error.

Comment (bhack) [reply tensorflow_addons/metrics/harmonic_mean.py:48]: Ok so `sample_weights` can be processed in the parent right?

Comment (rushabh-v) [reply tensorflow_addons/metrics/harmonic_mean.py:48]: Oh yes, got it!

File: tensorflow_addons/metrics/harmonic_mean.py (hunk 2 above)
Comment (rushabh-v) [reply tensorflow_addons/metrics/harmonic_mean.py:47]: Pushed it without `reciprocal_no_nan` and `masking`. Let me know if I need to change something.

File: tensorflow_addons/metrics/harmonic_mean.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(
+        self, name: str = "harmonic_mean", dtype: AcceptableDTypes = None, **kwargs
+    ):
+        super().__init__(name=name, dtype=dtype, **kwargs)
+
+    def update_state(self, values, sample_weight=None) -> None:
+        values = tf.cast(values, dtype=self.dtype)
+        super().update_state(tf.math.reciprocal(values), sample_weight)
+
+    def result(self) -> tf.Tensor:
```
Comment (bhack) [inline tensorflow_addons/metrics/harmonic_mean.py:49]: Is tf.Tensor ok here?

**Review by bhack — CHANGES_REQUESTED** [review]

File: tensorflow_addons/metrics/harmonic_mean.py (hunk 5 above)
Comment (rushabh-v) [reply tensorflow_addons/metrics/harmonic_mean.py:49]: numpy conversion was a mistake, `tf.keras.metrics.Mean` also returns a `Tensor`. Fixed it.

**Review by bhack — APPROVED** [review]
Now that is minimized I ask for a second opinion. @seanpmorgan let me know who can do a quick 2nd pass.

**Pull request merged**